### PR TITLE
Add engine_full column to system.databases

### PR DESCRIPTION
**Pull request by lingtaolf**
#42110
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add column `engine_full` to system table `databases` so that users can access whole engine definition of database via system tables.  

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by tavplubix** [comment]
Do we have a test for system.databases? Could you please add some if we don't?

**Comment by tavplubix** [comment]
Style Check failed, please take a look

File: tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+DROP DATABASE IF EXISTS replicated_database_test; 
+CREATE DATABASE IF NOT EXISTS replicated_database_test ENGINE = Replicated('some/path/replicated_database_test', 'shard_1', 'replica_1');
```
Comment (tavplubix) [inline tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.sql:2]:
```suggestion
CREATE DATABASE IF NOT EXISTS replicated_database_test ENGINE = Replicated('some/path/' || currentDatabase() || '/replicated_database_test', 'shard_1', 'replica_1');
```

Comment (tavplubix) [reply tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.sql:2]: Explanation: `clickhouse-test` generates unique database name for each test and it is used as current database for test. This change will guarantee that path is unique.

File: tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.reference [hunk 2]
```diff
@@ -0,0 +1 @@
+Replicated(\'some/path/replicated_database_test\', \'shard_1\', \'replica_1\')
```
Comment (tavplubix) [inline tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.reference:1]:
```suggestion
Replicated(\'some/path/default/replicated_database_test\', \'shard_1\', \'replica_1\')
```

Comment (tavplubix) [reply tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.reference:1]: (generated database name will be replaced with `default` in the test output before comparing it with the reference file)

File: src/Storages/System/StorageSystemDatabases.cpp [hunk 3]
```diff
@@ -47,7 +49,21 @@ void StorageSystemDatabases::fillData(MutableColumns & res_columns, ContextPtr c
...
+
+        ASTPtr ast = database->getCreateDatabaseQuery();
+        auto * ast_create = ast->as<ASTCreateQuery>();
+        String engine_full;
+
+        if (ast_create && ast_create->storage) {
+          engine_full = ast_create->storage->formatWithSecretsHidden();
+          static const char *const extra_head = " ENGINE = ";
+
+          if (startsWith(engine_full, extra_head))
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemDatabases.cpp:61]: These lines are redundant, formatted ASTStorage does not have `ENGINE = `

Comment (lingtaolf) [reply src/Storages/System/StorageSystemDatabases.cpp:61]: If delete these code, the result will be like below:
image
I'm not sure that's what you want.

**Comment by lingtaolf** [comment]
> Style Check failed, please take a look

done

File: src/Storages/System/StorageSystemDatabases.cpp (hunk 3 above)
Comment (tavplubix) [reply src/Storages/System/StorageSystemDatabases.cpp]: You're right, let's keep this code

File: tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.sql [hunk 4]
```diff
@@ -0,0 +1,5 @@
+DROP DATABASE IF EXISTS replicated_database_test; 
+CREATE DATABASE IF NOT EXISTS replicated_database_test ENGINE = Replicated('some/path/' || currentDatabase() || '/replicated_database_test', 'shard_1', 'replica_1');
```
Comment (tavplubix) [inline tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.sql:2]:
```suggestion
CREATE DATABASE IF NOT EXISTS replicated_database_test ENGINE = Replicated('some/path/' || currentDatabase() || '/replicated_database_test', 'shard_1', 'replica_1') SETTINGS max_broken_tables_ratio=1;
```
Let's also add some setting just to check that `SETTINGS` clause is printed as well.

**Review by tavplubix — APPROVED** [review]
LGTM, but failing tests have to be fixed (see Fast Test)

File: tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.sql (hunk 4 above)
Comment (lingtaolf) [reply tests/queries/0_stateless/02483_add_engine_full_column_to_system_databases.sql]: Updated

**Comment by tavplubix** [comment]
Fast test failed. There's a race condition between `DROP DATABASE` and `SELECT engine_full FROM system.databases`:
```
2022.11.25 13:52:57.185990 [ 15760 ] {} <Trace> DynamicQueryHandler: Request URI: /?query=CREATE+DATABASE+test_rhsp5cp6&database=system&connect_timeout=30&receive_timeout=30&send_timeout=30&http_connection_timeout=30&http_receive_timeout=30&http_send_timeout=30&log_comment=00762_date_comparsion.sql
2022.11.25 13:52:57.186198 [ 15760 ] {6c7e9ed5-aacc-471c-beea-3d7196062658} <Debug> executeQuery: (from [::1]:45038) (comment: 00762_date_comparsion.sql) CREATE DATABASE test_rhsp5cp6  (stage: Complete)
2022.11.25 13:52:57.186224 [ 15760 ] {6c7e9ed5-aacc-471c-beea-3d7196062658} <Trace> ContextAccess (default): Access granted: CREATE DATABASE ON test_rhsp5cp6.*
2022.11.25 13:52:57.186429 [ 15760 ] {6c7e9ed5-aacc-471c-beea-3d7196062658} <Information> DatabaseAtomic (test_rhsp5cp6): Metadata processed, database test_rhsp5cp6 has 0 tables and 0 dictionaries in total.
2022.11.25 13:52:57.186467 [ 15760 ] {6c7e9ed5-aacc-471c-beea-3d7196062658} <Information> DatabaseAtomic (test_rhsp5cp6): Starting up tables.
2022.11.25 13:52:57.245401 [ 15702 ] {} <Debug> TCPHandler: Connected ClickHouse client version 22.12.0, revision: 54460, database: test_rhsp5cp6, user: default.
2022.11.25 13:52:57.411816 [ 15629 ] {} <Trace> DynamicQueryHandler: Request URI: /?query=DROP+DATABASE+test_rhsp5cp6&database=system&connect_timeout=599&receive_timeout=599&send_timeout=599&http_connection_timeout=599&http_receive_timeout=599&http_send_timeout=599&log_comment=00762_date_comparsion.sql
2022.11.25 13:52:57.412050 [ 15629 ] {5bd116b9-52ed-427f-bde6-48d8d893c471} <Debug> executeQuery: (from [::1]:45462) (comment: 00762_date_comparsion.sql) DROP DATABASE test_rhsp5cp6  (stage: Complete)
2022.11.25 13:52:57.412078 [ 15629 ] {5bd116b9-52ed-427f-bde6-48d8d893c471} <Trace> ContextAccess (default): Access granted: DROP DATABASE ON test_rhsp5cp6.*
2022.11.25 13:52:57.444017 [ 15667 ] {b1ef0985-f100-412e-9827-652fe5e6b065} <Error> executeQuery: Code: 107. DB::ErrnoException: Cannot open file /fasttest-workspace/db-fasttest/metadata/test_rhsp5cp6.sql, errno: 2, strerror: No such file or directory. (FILE_DOESNT_EXIST) (version 22.12.1.1) (from [::1]:44506) (comment: 02423_ddl_for_opentelemetry.sh) (in query: select if(engine = 'Replicated', name, 'test_shard_localhost') from system.databases where name='test_7tybgaex'), Stack trace (when copying this message, always include the lines below):
2022.11.25 13:52:57.444165 [ 15667 ] {b1ef0985-f100-412e-9827-652fe5e6b065} <Error> TCPHandler: Code: 107. DB::ErrnoException: Cannot open file /fasttest-workspace/db-fasttest/metadata/test_rhsp5cp6.sql, errno: 2, strerror: No such file or directory. (FILE_DOESNT_EXIST), Stack trace (when copying this message, always include the lines below):
```

**Comment by evillique** [comment]
I think that a better solution to this task would include creating a separate method in IDatabase, for example: 
```
virtual String getEngineNameWithParameters() const
{
    return engine_name_with_parameters;
}
```
Where engine_name_with_parameters is a new protected String field.
And then, in the `DatabaseFactory::get()` function we would need to fill this field similar to `comment`.
 
@lingtaolf We found ourselves in a situation where we need to get this task done quite urgently, will you be able to do it? Or maybe we can help you with it?

**Comment by tavplubix** [comment]
Stateless tests (debug, s3 storage) [6/6] - https://github.com/ClickHouse/ClickHouse/pull/44202
Stress test (debug) - OOM in dmesg
Stress test (tsan) - https://github.com/ClickHouse/ClickHouse/pull/44197

**Pull request merged**